File: evopanel/src/lib/poll-loop-log.js
/** Throttled console logging for frontend poll loops (CPU troubleshooting).
 *
 * Default: OFF. Enable:
 *   localStorage.setItem('EVOFLOW_DEBUG_POLL', '1')
 */

const _lastTick = new Map()

function isPollDebugEnabled() {
  try {
    if (typeof localStorage === 'undefined') return false
    return localStorage.getItem('EVOFLOW_DEBUG_POLL') === '1'
  } catch {
    return false
  }
}

function ctxSuffix(ctx) {
  if (!ctx || typeof ctx !== 'object') return ''
  const parts = Object.entries(ctx)
    .filter(([, v]) => v != null && String(v).trim() !== '')
    .map(([k, v]) => `${k}=${v}`)
  return parts.length ? ` ${parts.join(' ')}` : ''
}

export function logPollLoopStart(name, ctx) {
  if (!isPollDebugEnabled()) return
  console.info(`[poll] ${name} 开始处理 loop${ctxSuffix(ctx)}`)
}

export function logPollLoopEnd(name, ctx) {
  if (!isPollDebugEnabled()) return
  console.info(`[poll] ${name} 结束处理 loop${ctxSuffix(ctx)}`)
}

export function logPollTick(name, key, intervalMs, ctx) {
  if (!isPollDebugEnabled()) return
  const cacheKey = `${name}:${key || '-'}`
  const now = Date.now()
  const last = _lastTick.get(cacheKey) || 0
  if (now - last < Math.max(1000, intervalMs || 30000)) return
  _lastTick.set(cacheKey, now)
  console.info(`[poll] ${name} 开始处理 tick${ctxSuffix({ key, ...ctx })}`)
}

File: backend/packages/harness/evoflow/observability/poll_loop_log.py
"""Throttled logging for long-running poll / SSE loops (CPU troubleshooting)."""

from __future__ import annotations

import logging
import time
from typing import Any

logger = logging.getLogger(__name__)

_last_tick: dict[str, float] = {}


def _ctx_suffix(ctx: dict[str, Any]) -> str:
    if not ctx:
        return ""
    parts = [f"{k}={v}" for k, v in ctx.items() if v is not None and str(v).strip() != ""]
    return (" " + " ".join(parts)) if parts else ""


def log_poll_loop_start(name: str, **ctx: Any) -> None:
    """Log once when a background poll loop / stream merger starts."""
    logger.info("[poll] %s 开始处理 loop%s", name, _ctx_suffix(ctx))


def log_poll_loop_end(name: str, **ctx: Any) -> None:
    """Log when a poll loop exits."""
    logger.info("[poll] %s 结束处理 loop%s", name, _ctx_suffix(ctx))


def log_poll_tick(
    name: str,
    *,
    key: str = "",
    interval_s: float = 30.0,
    level: int = logging.INFO,
    **ctx: Any,
) -> None:
    """Log a poll iteration at most once per ``interval_s`` per ``name:key``."""
    cache_key = f"{name}:{key or '-'}"
    now = time.monotonic()
    if now - _last_tick.get(cache_key, 0.0) < max(1.0, float(interval_s)):
        return
    _last_tick[cache_key] = now
    logger.log(level, "[poll] %s 开始处理 tick%s", name, _ctx_suffix(ctx))


def reset_poll_loop_log_cache() -> None:
    """Test helper."""
    _last_tick.clear()
